62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+
+#include "Hatch.h"
+using namespace nbl::hlsl;
+
+#include "common.hlsl"
+#include <nbl/builtin/hlsl/equations/cubic.hlsl>
+#include <nbl/builtin/hlsl/equations/quartic.hlsl>
+
+namespace hatchutils {
+    
+	// Intended to mitigate issues with NaNs and precision by falling back to using simpler functions when the higher roots are small enough
+	static std::array<double, 4> solveQuarticRoots(double a, double b, double c, double d, double e, double t_start, double t_end)
+	{
+		std::array<double, 4> t = { -1.0, -1.0, -1.0, -1.0 }; // only two candidates in range, ever
+
+		const double quadCoeffMag = std::max(std::abs(d), std::abs(e));
+		const double cubCoeffMag = std::max(std::abs(c), quadCoeffMag);
+		if (std::abs(a) > std::max(std::abs(b), cubCoeffMag) * QUARTIC_THRESHHOLD)
+		{
+			auto res = equations::Quartic<double>::construct(a, b, c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 4);
+		}
+		else if (abs(b) > quadCoeffMag)
+		{
+			auto res = equations::Cubic<double>::construct(b, c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 3);
+		}
+		else
+		{
+			auto res = equations::Quadratic<double>::construct(c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 2);
+		}
+
+		if (!(t[0] != t[1]))
+			t[0] = t[0] != t_start ? t_start : t_end;
+
+		//// TODO: check that this clamp works with t[i] as NaN
+		//for (auto i = 0; i < 2; i++)
+		//	t[i] = nbl::core::clamp(t[i], t_start, t_end);
+		//
+		//// fix up a fuckup where both roots are NaN or were on the same side of the valid integral
+		//
+		//// TODO: do some Halley or Householder method steps on t
+		////while ()
+		////{
+		////}
+
+		// neither t still not in range, your beziers don't intersect
+		return t;
+	}
+
+	// (only works for monotonic curves)
+	static float64_t2 getCurveRoot(double p0, double p1, double p2)
+	{
+		// https://people.csail.mit.edu/bkph/articles/Quadratics.pdf
+		double a = p0 - 2.0 * p1 + p2;
+		double b = 2.0 * (p1 - p0);
+		double c = p0;
+
+		// Solve linear equation
+		if (a == 0.0) return float64_t2(-c / b);
```

Comment:
why aren't you using the hlsl equations quadratic solve roots here?
also no need for that a==0.0 check here